tools/fdleak.py

```diff
@@ -0,0 +1,175 @@
+#
+# fdleak    Trace and display over-threshold fds to detect FD leaks process.
+#           All FD allocate will be traced, if the threshold is exceeded or 
+#           EMFILE occurs, then will be recorded by the monitoring program.
+#
+#           `In UNIX, everything is a file`, file leakage problems may 
+#           occur when most resources are used improperly, eg:
+#             file, socket, eventfd, pipe, ashmem, dmabuf, syncfence ...
+#
+#           It's a light and easy to use ebpf probe fd alloc and monitor 
+#           over-threshold fds process, collects stack and list open files.
+#
+# USAGE: fdleak [-h] [-p PID] [-w WARN] [--dumpstack] [--lsof] [-a]
+#
+# Licensed under the Apache License, Version 2.0 (the "License")
+# Copyright (c) 2021 Vachel Yang.
+# 11-Oct-2021    Vachel.Yang    Created this.
+
+from bcc import BPF
+from bcc.utils import printb
+from datetime import datetime
+import argparse,sys,os
+
+# arguments
+examples = """examples:
+    ./fdleak                # trace all process alloc_fd()
+    ./fdleak -p 181         # only trace PID 181
+    ./fdleak -w 900         # set warning threshold 900
+    ./fdleak --dumpstack    # show alloc fd stack
+    ./fdleak --lsof         # show open files
+    ./fdleak -a             # show all debug info
+"""
+
+parser = argparse.ArgumentParser(
+    description="Trace FD leak",
+    formatter_class=argparse.RawDescriptionHelpFormatter,
+    epilog=examples)
+parser.add_argument("-p", "--pid",
+                    help="trace this PID only")
+parser.add_argument("-w", "--warn",
+                    help="set warning threshold,default is 819(1024*0.8)")
+parser.add_argument("--dumpstack",default=False, action="store_false", dest="dumpstack",
+                    help="show alloc fd stack")
+parser.add_argument("--lsof",default=False, action="store_false", dest="lsof",
+                    help="list of open files")
+
+parser.add_argument("-a", "--showall",default=False, action="store_false",
+                    help="show all debug info")
+args = parser.parse_args()
+
+if args.showall:
+    args.lsof=True
+    args.dumpstack=True
+
+bpf_text = """
+#include <uapi/linux/ptrace.h>
+#include <bcc/proto.h>
+#include <asm-generic/errno-base.h>
+#include <linux/fdtable.h>
+#include <linux/sched.h>
+
+BPF_PERF_OUTPUT(events);
+// 8k buf
+BPF_STACK_TRACE(kernel_stack, 8192);
+// 8k buf
+BPF_STACK_TRACE(user_stack, 8192);
+
+struct data_t {
+    char taskname[TASK_COMM_LEN];
+    int pid;
+    int fds;
+    int stack_user;
+    int stack_kernel;
+};
+
+inline static void submit_to_user(int fds,struct pt_regs *ctx){
+    struct data_t data = {0};
+    data.pid = bpf_get_current_pid_tgid();
+    data.fds = fds;
+    bpf_get_current_comm(&data.taskname, sizeof(data.taskname));
+    data.stack_user = user_stack.get_stackid(ctx, BPF_F_USER_STACK | BPF_F_REUSE_STACKID);
+    data.stack_kernel = kernel_stack.get_stackid(ctx, 0);
+    events.perf_submit(ctx, &data, sizeof(data));
```

Comment:
Shouldn't use perf submit, because a lot of normal FD operations for production case.

Reply:
This problem has also been fixed, replace  BPF_PERF_ARRAY with BPF_HASH. Thanks for pointing out the mistake.